(WIP) Implement fasttrap for mips

nicomazz opened the pull request:
Most of the code is taken from powerpc, with sensible changes

nicomazz committed:
Implement fasttrap for mips - beginning

The port for mips has started. However, it's at the very beginning.
Most of the code comes from the powerpc implementation.

nicomazz committed:
Add parameter_dtrace_test

The reason it so clearly see the disassembly in function calls

nicomazz committed:
Merge branch 'master' into dtrace_mips_pid

nicomazz committed:
Generate CTF data from .dynsym section for mips

For mips, the CTF section is generated from the symbols in `.dynsym`.
`.dymsym` only contains a subset of the `.symtab` symbols, needed for
dynamic linking. This flag is needed because the `.symtab` section is not
loaded at boot time, and its address is not available anywhere without a
proper bootloader.
The problem can be solved looking at |sys/mips/mips/elf_trampoline.c|.

nicomazz committed:
Fix kernel pathname when loading CTF for mips

DTrace was failing to load the kernel ELF because its path was only "kernel"
instead of "/boot/kernel/kernel".

nicomazz committed:
Add todo for DTrace fbt type name

nicomazz committed:
Fix CTF pointer overrun (in userland)

Cheri catched this error. The problem arose when p = "s", and lp->ctl_prefix = "struct".
strncmp("s","struct",1) returns 0, because it checks only the first char.
The following line was advancing `p`(="s") by 6, going over the terminator.
(the value of p can be user controlled)

nicomazz committed:
Enable CTF generation for kernel

CDDL was set as a broken option for mips. It is not so broken anymore.

nicomazz committed:
Fix CTFFLAGS in kern.post.mk

nicomazz committed:
Fix CTFFLAGS in kern.post.mk

Changing CTFFLAGS directly didn't work because it was used in other places that didn't require the -s flag

nicomazz committed:
Comment MODULES_OVERRIDE for dtrace config

nicomazz committed:
Add todo for fbt_provide_module_function

nicomazz committed:
Compile fasttrap for mips into kernel

The necessary glue has been added to compile fasttrap into the kernel, and not as a dynamically loadable module (which is still possible)/\.
Also, systrace_freebsd32 is not compiled if we are in mips. There where too many errors.

nicomazz committed:
Fix several mips fasttrap errors

I doesn't work yet, but at least it compiles fine.

nicomazz committed:
Fix compilation issues for mips fasttrap

nicomazz committed:
Fix kldxref cheri-related issue

`dest` is a capability with a len L. The code was advancing this pointer by (way) more than L, invalidating the TAG, and then make it step back. However, at the end the TAG was still invalid. In this way the offset is precomputed.

nicomazz committed:
Fix ioctl special case with _IO

There are ioctls defined as _IO (void, without input and output), that actually need the parameter, even if their IOCPARAM_LEN is zero.
An example is for fasttrap: the ioctl FASTTRAPIOC_MAKEPROBE needs a parameter, buy is defined as:
`#define	FASTTRAPIOC_MAKEPROBE	_IO('f', 3)`
The reason it is defined like this is to skip the `user_ioctl` copyin, and handle the copyin logic later in fasttrap_ioctl. This because the len of the copyin depends by a parameter.
For CHERI, the data parameter of ioctls was set to NULL if IOCPARAM_LEN returned zero.

brooksdavis commented on lib/libc/sys/ioctl.c:
```diff
@@ -46,6 +46,7 @@ __FBSDID("$FreeBSD$");
 #include <sys/types.h>
 #include <stdarg.h>
 #include <stddef.h>
+#include <cheri/cheric.h>
```
#ifdef...

brooksdavis commented on lib/libc/sys/ioctl.c:
```diff
@@ -88,9 +89,13 @@ _ioctl(int fd, unsigned long com, ...)
-		va_end(ap);
 	} else {
-		data = NULL;
+		if (cheri_getlen((void*)ap) == 16)
+			data = va_arg(ap, void *);
+		else
+			data = NULL;
```
I think we want to handle the sane cases.  I believe that's int (char is promoted to int IIRC), long, and void *. I've chosen int##_t because I believe we want sign extension here but I could be wrong.

```suggestion
#ifdef __CHERI_PURE_CAPABILITY__
		if (cheri_getlen((void*)ap) == sizeof(void *))
			data = va_arg(ap, void *);
		else if (cheri_getlen((void*)ap) == sizeof(uint64_t))
			data = (void *)(intptr_t)va_arg(ap,  int64_t);
		else if (cheri_getlen((void*)ap) == sizeof(int32_t))
			data = (void *)(intptr_t)va_arg(ap, int32_t);
		else
			data = NULL;
#else
		/* Assume this was passed in a register. */
		data = (void *)(intptr_t)va_arg(ap, register_t);
#endif
```

nicomazz committed:
Handle other cases in ioctl.c when size is zero

nicomazz committed:
Enable Ptrace getregs for CheriABI (not PT_SETCAPREG)

nicomazz committed:
Add debug printf for unimplemented mips feature

nicomazz committed:
Fix fasttrap mips PC handling

Still, it doesn't work, yet

nicomazz committed:
Fix casts for addresses in proc debug printfs

nicomazz committed:
Fix proc_regs to handle Cheri PC offset

in the normal registers gathered from PT_GETREGS the pc is only the offset.
To have the complete address we have to add the base of the PCC.
To make it faster, it is possible to get all directly from the PCC. However, we still need to fetch the all the normal registers, so the performance for the set case are the same.

nicomazz committed:
Calculate offset from first file mapping

This patch is from markj.
The symbols in the .symtab are offsets from the first mapping of each file. The offset was calculated in a wrong way. As a consequence, libproc was unable to find the correspondence between symbols and addresses.
Now all libproc tests are fixed.

nicomazz committed:
Merge from dtrace_support

nicomazz committed:
Fix LD_ELF name in proc_test

The name in the process mappings is always ld-elf.
With ld-cheri-elf it wasn't finding it.

nicomazz committed:
Add KTR to CHERI128_DTRACE

nicomazz committed:
Add __FreeBSD__ to CMakeLists.txt

nicomazz committed:
Add new break instruction for DTrace pid provider

nicomazz committed:
Fix fasttrap_isa.c for mips

nicomazz committed:
Add DTrace pid provider breakpoints case to trap.c

nicomazz committed:
Merge from dev

nicomazz committed:
Remove parameter test

nicomazz committed:
Revert useless changes

nicomazz committed:
Fix proc_regs

nicomazz committed:
Make fasttrap optional on `dtrace_fasttrap`

nicomazz committed:
Avoid fetching normal registers when we request the PC for mips.

nicomazz committed:
Remove KTR from CHERI128_DTRACE

nicomazz committed:
Disable breakpoints while probes are created in PID provider

nicomazz committed:
Fix uintprt in fasttrap

nicomazz committed:
Make pid provider work in mips singlestepping

nicomazz committed:
Add check for breakpoints in libproc mips

nicomazz committed:
Remove wrong endif

nicomazz committed:
Implement pid create offset probe

nicomazz committed:
Implement return probe for mips pid

nicomazz committed:
Fix dt_proc breakpoint match

In MIPS, also fasttrap can set breakpoints.
To be sure that the matching is from a breakpoint set by dt_proc, it is firstly verified that (the bp) is active).

nicomazz committed:
Implement Fasttrap for mips

Behaviour with branches has been fixed here

nicomazz committed:
Fix check for mips

nicomazz committed:
Add Branch cheri opcodes and fix MipsEmulateBranch

nicomazz committed:
Fix log, remove ioctl

nicomazz committed:
Fix semicolon

nicomazz committed:
Merge branch 'dtrace_support' into dtrace_mips_pid

nicomazz committed:
Various fixes to fasttrap for the noncap case

nicomazz committed:
Merge from dev

nicomazz committed:
Fix ustack for dtrace

Now it's possible to generate stack traces for a purecap userspace.

I assume that:
1. The stack pointer (c11) is modified at the beginning of the function in either

`cincoffsetimm   $c11,$c11,-X`
or
`daddiu  at,zero,-X
cincoffset      $c11,$c11,at`

2. The return pointer is saved into the stack with `csc     $c17,zero,X($c11)`

In some case, this fails and `dtrace: ERROR: proc_addr2map failed to resolve ...` is printed, so there might be something wrong with the assumptions I made.

To test it: `dtrace -n "profile-1ms {	@[ustack()] = count(); }" -c date`
...
syslogd`fprintlog_first+0xa4
syslogd`logmsg+0x4e8
dtrace: ERROR: proc_addr2map failed to resolve 0x100000000
  7
libc.so.7`memcpy+0x30
libc.so.7`__sfvwrite+0x2b8
libc.so.7`__vfprintf+0x3dbc
libc.so.7`snprintf+0xf0
  8

libc.so.7`__vfprintf+0x3b8
libc.so.7`snprintf+0xf0
  8

libc.so.7`memcpy+0x134
libc.so.7`__sfvwrite+0x2b8
libc.so.7`__vfprintf+0x3dbc
libc.so.7`snprintf+0xf0
 14

nicomazz committed:
Remove debug printfs from ustack

nicomazz committed:
Add some cheri MIPS opcode

nicomazz committed:
Fix dtrace.h ifdef

nicomazz committed:
Fix fasttrap printf

nicomazz committed:
Fix sys_ptrace for Cheri

In this way, fasttrap works also without cheriabi.
Most of the changes are to make sys_ptrace similar to the one in cheriabi_ptrace.c

nicomazz committed:
Merge with DTrace_support

nicomazz committed:
Fix sys_ptrace

Now the default case is a purecap process, and many ifdefs with COMPAT_FREEBSD64 have been added

nicomazz committed:
Revert a few changes in sys_ptrace

nicomazz committed:
Address review comments

nicomazz committed:
Remove dtraceall from CHERI128_DTRACE

nicomazz committed:
Fix compat64 issue in dtrace_ioctl

nicomazz committed:
Add fasttrap to CHERI128_DTRACE
